tests/integrations/mcs/keyspace/tso_keyspace_group_test.go

```diff
@@ -495,3 +551,163 @@ func (suite *keyspaceGroupTestSuite) trySetNodesForKeyspaceGroup(re *require.Ass
+// become temporarily unavailable and then recover, the client should NOT fallback to
+// the legacy path (group 0), but should wait and successfully get TSO after nodes restart.
+//
+// Test scenario:
+// 1. Setup: Start 2 TSO nodes and create keyspace group 1, client gets initial TSO
+// 2. Close all TSO nodes to simulate total TSO microservice failure
+// 3. Wait until all TSO nodes are deregistered (getTSOServerURLs returns empty)
+// 4. Enable failpoints: assertNotReachLegacyPath (panic if fallback) and extend timeout
+// 5. Start async GetTS call (will wait for TSO service to recover)
+// 6. Restart one TSO node while GetTS is waiting
+// 7. Verify GetTS succeeds after node restart (assertNotReachLegacyPath ensures no fallback)
+func (suite *keyspaceGroupTestSuite) TestUpdateMemberWhenRecovery() {
+	re := suite.Require()
+
+	// Test configuration constants
+	const (
+		// TSO dispatcher timeout extension to prevent timeout
+		tsoDispatcherTimeoutSeconds = 30
+		// Time to wait for GetTS to start
+		waitForGetTSStart = 2 * time.Second
+	)
+
+	// Step 1: Setup - Create 2 TSO nodes and client, get initial TSO
+	setup := suite.setupTSONodesAndClient(re, 2, 1)
+	defer func() {
+		for _, cleanup := range setup.cleanups {
+			cleanup()
+		}
+	}()
+	defer setup.client.Close()
+
+	nodes := setup.nodes
+	client := setup.client
+	firstNodeAddr := setup.firstNodeAddr
+
+	// Step 2: Close all TSO nodes to simulate total TSO microservice failure
+	// Step 3: Wait until all TSO nodes are deregistered
+	suite.closeAllTSONodesAndWait(re, nodes)
+
+	// Step 4: Enable failpoints for verification and timeout extension
+	// assertNotReachLegacyPath: ensures we don't fallback to legacy path (will panic if we do)
+	re.NoError(failpoint.Enable("github.com/tikv/pd/client/servicediscovery/assertNotReachLegacyPath", "return(true)"))
+	defer func() {
+		re.NoError(failpoint.Disable("github.com/tikv/pd/client/servicediscovery/assertNotReachLegacyPath"))
+	}()
+	// tsoDispatcherExtendTimeout: prevents GetTS from timing out while waiting for node restart
+	re.NoError(failpoint.Enable("github.com/tikv/pd/client/client/clients/tso/tsoDispatcherExtendTimeout",
+		fmt.Sprintf("return(%d)", tsoDispatcherTimeoutSeconds)))
+	defer func() {
+		re.NoError(failpoint.Disable("github.com/tikv/pd/client/client/clients/tso/tsoDispatcherExtendTimeout"))
+	}()
+
+	// Step 5: Start async GetTS call - it will wait for TSO service to recover
+	type getTSResult struct {
+		physicalTS int64
+		logicalTS  int64
+		err        error
+	}
+	resultCh := make(chan getTSResult, 1)
+
+	getTSStarted := make(chan struct{})
+	go func() {
+		close(getTSStarted) // Signal that GetTS call has started
+		physicalTS, logicalTS, getErr := client.GetTS(suite.ctx)
+		resultCh <- getTSResult{physicalTS: physicalTS, logicalTS: logicalTS, err: getErr}
+	}()
+
+	// Wait for GetTS goroutine to start
+	<-getTSStarted
```

Comment:
Can it pass in the master branch for this test? I gause this rquest will send to the pd server, so the result will not wait.